stm32_common: Add USB OTG FS/HS usbdev peripheral driver

bergzand opened the pull request:
### Contribution description

This PR adds an USBdev driver for the ST (synopsys). It is used with the STM32F2, STM32F4, STM32F7 and STM32H7 devices. The STM32F1 and STM32F3 devices unfortunately use a different device type.

The goal is to support the both the full speed and the high speed version in a single driver, the differences between the two are minimal and drivers from [other](https://github.com/libopencm3/libopencm3/blob/master/lib/usb/usb_dwc_common.c) [project](https://github.com/STMicroelectronics/STM32CubeF4/blob/master/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_ll_usb.c) are able to combine the two instances too.

The peripheral itself has a 'Core ID' register containing a version number-like value. So far I've identified 2 different iterations for both the FS and the HS devices. They have slightly different register flags and behave slightly different in some cases. The crude method to detect the version is to check [if certain register flag definitions are defined](https://github.com/RIOT-OS/RIOT/compare/master...bergzand:wip/stusbdev?expand=1#diff-1ccd6f95490bc49e1c979bf9e4ca3590R51-R57).
It is assumed that when both the FS and the HS peripheral is available on a single device, both are of the same version.

- CID `0x1x00`: 4 FS endpoints, 6 HS endpoints (f401, f411…)
- CID `0x2x00`: 6 FS endpoints, 9 HS endpoints (f446, f7xx…)

### Testing procedure

- `examples/usbus_minimal` should work when adding `xtimer`.
- `tests/usbus_cdc_ecm` should work with #12533 cherry-picked.
- `tests/usbus_cdc_acm` should work with #12536 cherry-picked.

### Issues/PRs references

One of the remaining issues with the driver is that it requires xtimer (or a different solution) to sleep between one of the register modifications.

depends on ~~#12534~~, ~~#12533~~, ~~#12536~~ and ~~#12579~~

### TODO:

- [x] Add xtimer dependency when `periph_usbdev` is used.
- [x] Add usbdev feature to non-nucleo boards
- [x] Test with the HS peripheral in FS mode
- [x] Add DMA support for the HS peripheral
- [x] Add support for the non-nucleo boards
- [x] Migrate DMA enable/disable to a `#define`
- [x] ~~Test isochronous transfers~~

### Note

@chrysn, @basilfx: FYI The EFM32 devices seem to use an identical IP block from Synopsys, only the serialization phy seems to be different.

bergzand commented:
Almost done here, I've tested this myself on the nucleo-f401re and the nucleo-f446re with a jumper wired USB cable. Both CID versions should work and both the FS and the HS peripheral work (in FS mode with the built-in phy). With the HS phy, both DMA and non-DMA is functional.

I have not yet tested isochronous transfers.

bergzand commented:
Ready for review! I won't be force pushing anymore to this branch unless instructed otherwise.

I'm going to test isochronous transfers at some point, but I won't consider it blocking for this PR as it is not used within RIOT-provided USB functionality at the moment.

Still waiting for #12579 though. :)

bergzand commented:
I'm carefully going to claim that isochronous transfers work. I have a small test setup where I can at least get the events from the transfer and receive events for the transfers. I also receive correct transfer lengths, I'm only haven't tested the content of the transfer yet.

bergzand commented:
Rebased now that #12579 is merged, no longer waiting for other PRs

bergzand commented:
> I'm carefully going to claim that isochronous transfers work.

I was wrong, further testing showed that the application is responsible for toggling the data0/data1 flag for isochronous transfers.

benpicco commented:
I just skimmed over the code so far, but looks good.
However, when I try with `stm32f429i-disc1` I have no luck.
All the USB examples run, but nothing shows up in `dmesg` when I plug the device.

bergzand commented:
>  However, when I try with `stm32f429i-disc1` I have no luck.

Looks like ST switched to the HS peripheral for this board. I've added ad config and replaced the include in the board `periph_conf.h`

benpicco commented:
With the last update I see something when flashing `usbus_cdc_acm_stdio` on `stm32f429i-disc1`
```
[  136.661753] usb 1-1.5: new full-speed USB device number 4 using ehci-pci
[  156.374672] usb 1-1.6: new full-speed USB device number 5 using ehci-pci
[  171.723115] usb 1-1.6: device descriptor read/64, error -110
[  187.339431] usb 1-1.6: device descriptor read/64, error -110
[  187.527416] usb 1-1.6: new full-speed USB device number 6 using ehci-pci
[  202.955650] usb 1-1.6: device descriptor read/64, error -11
```

benpicco commented:
I see the same on `WeAct-f411ce` with `cfg_usb_otg_fs.h`:

```
[  747.216730] usb 1-1.5: new full-speed USB device number 18 using ehci-pci
[  762.576886] usb 1-1.5: device descriptor read/64, error -110
[  778.193080] usb 1-1.5: device descriptor read/64, error -110
[  778.381043] usb 1-1.5: new full-speed USB device number 19 using ehci-pci
[  793.809195] usb 1-1.5: device descriptor read/64, error -110
```

bergzand commented:
Thanks for testing, I'll try to reproduce tomorrow, see what I can find.

bergzand commented:
I've been unable to reproduce this so far on my `nucleo-f401re`, using `tests/usbus_cdc_acm_stdio`:
```
[115064.416737] usb 3-9.2.2: new full-speed USB device number 24 using xhci_hcd
[115064.496271] usb 3-9.2.2: New USB device found, idVendor=1209, idProduct=0001, bcdDevice= 0.00
[115064.496273] usb 3-9.2.2: New USB device strings: Mfr=3, Product=2, SerialNumber=0
[115064.496274] usb 3-9.2.2: Product: USB device
[115064.496274] usb 3-9.2.2: Manufacturer: RIOT-os.org
[115064.504521] cdc_acm 3-9.2.2:1.0: ttyACM0: USB ACM device
```

benpicco commented on cpu/stm32_common/periph/usbdev.c:
```diff
@@ -0,0 +1,1168 @@
+
+static bool _uses_dma(const stm32_usb_otg_fshs_config_t *config)
+{
+#if STM32_USB_OTG_HS_USE_DMA
```
```suggestion
#if STM32_USB_OTG_HS_ENABLED && STM32_USB_OTG_HS_USE_DMA
```

This will save you a lot of `#ifdef`s later on.

benpicco commented on cpu/stm32_common/periph/usbdev.c:
```diff
@@ -0,0 +1,1168 @@
+        if (_uses_dma(conf)) {
+            _in_regs(conf, ep->num)->DIEPDMA = (uint32_t)ep->buf;
+        }
+#endif  /* STM32_USB_OTG_HS_ENABLED */
```
You can drop the `#ifdef` then.

benpicco commented on cpu/stm32_common/periph/usbdev.c:
```diff
@@ -0,0 +1,1168 @@
+        if (len > 0
+#ifdef STM32_USB_OTG_HS_ENABLED
+            && !_uses_dma(conf)
+#endif  /* STM32_USB_OTG_HS_ENABLED */
```
and here

benpicco commented on cpu/stm32_common/periph/usbdev.c:
```diff
@@ -0,0 +1,1168 @@
+        if (_uses_dma(conf)) {
+            _out_regs(conf, ep->num)->DOEPDMA = (uint32_t)ep->buf;
+        }
+#endif  /* STM32_USB_OTG_HS_ENABLED */
```
here too

benpicco commented on cpu/stm32_common/periph/usbdev.c:
```diff
@@ -0,0 +1,1168 @@
+            ep->num
+#ifdef STM32_USB_OTG_HS_ENABLED
+            && !_uses_dma(conf)
+#endif  /* STM32_USB_OTG_HS_ENABLED */
```
and here.

benpicco commented on cpu/stm32_common/periph/usbdev.c:
```diff
@@ -0,0 +1,1168 @@
+            if (_uses_dma(conf)) {
+                usbdev->usbdev.epcb(ep, USBDEV_EVENT_TR_COMPLETE);
+            }
+#endif      /* STM32_USB_OTG_HS_ENABLED */
```
dito.

benpicco commented on cpu/stm32_common/periph/usbdev.c:
```diff
@@ -0,0 +1,1168 @@
+
+#ifdef STM32_USB_OTG_HS_ENABLED
+    if (_uses_dma(conf)) {
+        _global_regs(usbdev->config)->GAHBCFG |=
```
same here.

fjmolinas commented:
> One of the remaining issues with the driver is that it requires xtimer (or a different solution) to sleep between one of the register modifications.

You would like to get rid of xtimer? or get rid of the reason why the sleep is needed?

aabadie commented on cpu/stm32_common/periph/usbdev.c:
```diff
@@ -0,0 +1,1168 @@
+    _global_regs(conf)->GUSBCFG |= USB_OTG_GUSBCFG_FDMOD;
+    /* Spinlock to prevent a context switch here, needs a delay of 25 ms when
+     * force switching mode */
+    xtimer_spin(xtimer_ticks_from_usec(25 * US_PER_MS));
```
You can get rid of the xtimer dependency by simply use a while loop and the value of CLOCK_CORECLOCK to spin during 25ms.

benpicco replied:
Only if your CPU does one loop iteration per clock cycle.

aabadie commented:
ping @bergzand!

aabadie commented:
@bergzand, I'd like to see this PR go in the release. Any chance you could address/answer the remaining comments ? Thanks!

benpicco commented:
My changes have been addressed :wink: 
Please squash (& rebase)

bergzand commented:
Squashed! @benpicco I might have accidentally removed your authorship in the squash. Is that a problem for you?

aabadie commented:
You need to rebase. What is the decision regarding the use xtimer in the peripheral implementation ?

bergzand commented:
> You need to rebase.

Rebased!

>  What is the decision regarding the use xtimer in the peripheral implementation ?

How about we leave it as is now and see if we want to swap it for something else if this causes issues? I think the current code is okay, it is just that I don't like using xtimer in a low level peripheral.

benpicco commented on boards/common/stm32/include/cfg_usb_otg_hs_fs.h:
```diff
@@ -0,0 +1,62 @@
+}
+#endif
+
+#endif /* CFG_USB_OTG_FS_H */
```
```suggestion
#endif /* CFG_USB_OTG_HS_FS_H */
```

bergzand replied:
amended!

dylad commented on cpu/stm32_common/include/usbdev_stm32.h:
```diff
@@ -0,0 +1,136 @@
+ */
+#ifndef STM32_USB_OTG_HS_USE_DMA
+#ifdef STM32_USB_OTG_CID_1x
+#define STM32_USB_OTG_HS_USE_DMA        (0)     /* FIXME */
```
Why FIXME ?

bergzand replied:
Because it should be possible to use the built-in DMA with this peripheral type, but somehow it doesn't work.

But I guess you want me to add this to that comment :)

dylad replied:
You read in my mind :)

bergzand replied:
added

dylad commented on cpu/stm32_common/periph/usbdev.c:
```diff
@@ -0,0 +1,1151 @@
+
+/* Some device families forgot to define the FS device FIFO size */
+#ifndef USB_OTG_FS_TOTAL_FIFO_SIZE
+#define USB_OTG_FS_TOTAL_FIFO_SIZE      (1280U)
```
1280 ? Isn't a weird value ?

bergzand replied:
[I don't think it is that weird](https://github.com/RIOT-OS/RIOT/blob/master/cpu/stm32f4/include/vendor/stm32f401xe.h#L8493)

dylad replied:
How about use the vendor define then?

bergzand replied:
Some of the ST vendor headers (for the parts that do have this peripheral) are missing this define.

dylad replied:
But are we 100% sure, this value will remain the same across families ?

bergzand replied:
Probably not, but I don't really see a different solution. I have already wrapped it in an `#ifndef` to allow the vendor headers to override the value with their own. I think that's the best we can do for now.

bergzand replied:
The OTG fs peripheral is present, but the FIFO size define is missing in the following device vendor headers:
```
cpu/stm32f7/include/vendor/stm32f722xx.h
cpu/stm32f7/include/vendor/stm32f723xx.h
cpu/stm32f7/include/vendor/stm32f746xx.h
cpu/stm32f7/include/vendor/stm32f767xx.h
cpu/stm32f7/include/vendor/stm32f769xx.h
cpu/stm32l4/include/vendor/stm32l475xx.h
cpu/stm32l4/include/vendor/stm32l476xx.h
cpu/stm32l4/include/vendor/stm32l496xx.h
cpu/stm32l4/include/vendor/stm32l4r5xx.h
```

dylad commented on cpu/stm32_common/periph/usbdev.c:
```diff
@@ -0,0 +1,1151 @@
+}
+
+static __O uint32_t *_tx_fifo(const stm32_usb_otg_fshs_config_t *conf,
+                              size_t num)
```
Why is there an extra param (size_t num) for tx_fifo but not for rx_fifo ?

bergzand replied:
Because there are multiple TX FIFOs (for IN transfers), but only a single shared RX FIFO (for OUT transfers).

dylad commented on cpu/stm32_common/periph/usbdev.c:
```diff
@@ -0,0 +1,1151 @@
+ * @param   conf      usbdev context
+ * @param   fifo_num    fifo number to reset, 0x10 for all fifos
+ */
+static void _flush_fifo(const stm32_usb_otg_fshs_config_t *conf,
```
```suggestion
static void _flush_tx_fifo(const stm32_usb_otg_fshs_config_t *conf,
```
For consistency

bergzand replied:
changed

dylad commented on cpu/stm32_common/periph/usbdev.c:
```diff
@@ -0,0 +1,1151 @@
+            _ep_set_stall(ep, *(usbopt_enable_t *)value);
+            res = sizeof(usbopt_enable_t);
+            break;
+        default:
```
Please add a DEBUG statement here.

bergzand replied:
I've added the DEBUG statement and cleaned up the other debug statements where relevant.

bergzand commented:
> This driver seems far more complicate than the previous one...

I know :sob:

dylad commented on cpu/stm32_common/periph/usbdev.c:
```diff
@@ -0,0 +1,1151 @@
+        /* disable endpoint and set NAK */
+        _out_regs(conf, num)->DOEPCTL = USB_OTG_DOEPCTL_EPDIS | USB_OTG_DOEPCTL_SNAK;
+        /* Wait for the disable to take effect */
+        while (_in_regs(conf, num)->DIEPCTL & USB_OTG_DOEPCTL_EPDIS) {}
```
```suggestion
        while (_out_regs(conf, num)->DIEPCTL & USB_OTG_DOEPCTL_EPDIS) {}
```

bergzand replied:
Good catch!

dylad replied:
I would rather say
```suggestion
        while (_out_regs(conf, num)->DOEPCTL & USB_OTG_DOEPCTL_EPDIS) {}
```

bergzand replied:
Fixed!

dylad commented on cpu/stm32_common/periph/usbdev.c:
```diff
@@ -0,0 +1,1151 @@
+    _flush_fifo(conf, 0x10);
+
+    /* Values from the reference manual tables on TRDT */
+    uint8_t trdt = conf->type == STM32_USB_OTG_FS ? 0x06 : 0x09;
```
Is it possible to document this value or add a reference to the doc ?

bergzand replied:
I think I decided against a reference because the exact table number differ between the different MCU reference manuals. I'll try to clarify it a bit

dylad commented:
@bergzand
Please squash, I'll give it one last try tonight as soon as I have access to my hardware.

aabadie commented on cpu/stm32_common/periph/usbdev.c:
```diff
@@ -0,0 +1,1150 @@
+ * @author  Koen Zandberg <koen@bergzand.net>
+ * @}
+ */
+
```
```suggestion
#define USB_H_USER_IS_RIOT_INTERNAL

```

bergzand replied:
Added in b440a0b

bergzand commented:
I've separated out the commits which enable the peripheral on the L4 boards. I suspect a clock configuration issue, but this is only a gut feeling.

aabadie commented on cpu/stm32_common/Makefile.dep:
```diff
@@ -3,3 +3,7 @@ USEMODULE += pm_layered
 USEMODULE += stm32_common stm32_common_periph
+
+ifneq (,$(filter periph_usbdev,$(FEATURES_USED)))
+	USEMODULE += xtimer
```
```suggestion
  USEMODULE += xtimer
```

Looks like there's a tab here

bergzand replied:
:astonished: Fixed!

bergzand commented:
Squashed after offline approval from @aabadie.

aabadie commented:
All green, let's go !

bergzand commented:
Thanks!

dylad commented:
Good job with this PR @bergzand !
